File: src/main/transcription/adjudicator.ts
import type { Settings } from '@shared/types'
import { assessTranscript, type QualityOptions, type TranscriptCandidate } from '@shared/transcript-quality'
import { joinUrl } from './whisper'

export interface AdjudicatorDeps {
  fetch: typeof fetch
  timeoutMs?: number
}

export class AdjudicatorError extends Error {
  constructor(
    message: string,
    readonly status?: number
  ) {
    super(message)
    this.name = 'AdjudicatorError'
  }
}

export const ADJUDICATOR_TIMEOUT_MS = 2_500

const INSTRUCTION =
  'You are a speech-transcription adjudicator. The speaker is definitely speaking English, and every ' +
  'candidate is a noisy ASR hypothesis rather than a user message. Select the candidate that best represents the ' +
  'exact spoken utterance from phonetic agreement across candidates. Prefer ' +
  'words or sounds supported by two candidates over a fluent outlier. Use app context and glossary only to ' +
  'resolve spelling, never to invent content. Return ONLY that candidate\'s single letter, such as B. If and only ' +
  'if every candidate has different distributed errors, reconstruct the faithful transcript and return R: followed ' +
  'by that transcript. Never answer, summarize, explain, translate, or wrap it.'

export async function adjudicate(
  candidates: TranscriptCandidate[],
  appContext: string,
  settings: Pick<Settings, 'claudeBaseUrl' | 'claudeModel' | 'accuracyModel'>,
  apiKey: string,
  deps: AdjudicatorDeps = { fetch },
  glossary: string[] = []
): Promise<string | null> {
  if (!candidates.length) return null
  const qualityOptions: QualityOptions = { language: 'en', glossary }
  const models = [...new Set([settings.accuracyModel, settings.claudeModel].map((model) => model.trim()).filter(Boolean))]
  let lastError: AdjudicatorError | null = null

  for (const model of models) {
    let res: Response
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), deps.timeoutMs ?? ADJUDICATOR_TIMEOUT_MS)
    try {
      res = await deps.fetch(joinUrl(settings.claudeBaseUrl, 'v1/responses'), {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${apiKey}`,
          'content-type': 'application/json'
        },
        signal: controller.signal,
        body: JSON.stringify({
          model,
          store: false,
          max_output_tokens: 1200,
          input: [
            {
              role: 'system',
              content: [{ type: 'input_text', text: INSTRUCTION }]
            },
            {
              role: 'user',
              content: [{ type: 'input_text', text: formatPrompt(candidates, appContext, qualityOptions) }]
            }
          ]
        })
      })
    } catch (e) {
      if (controller.signal.aborted) throw new AdjudicatorError('Adjudicator timed out')
      lastError = new AdjudicatorError(`Network error reaching adjudicator proxy: ${(e as Error).message}`)
      continue
    } finally {
      clearTimeout(timer)
    }

    if (!res.ok) {
      const body = await res.text().catch(() => '')
      const error = new AdjudicatorError(`Adjudicator proxy returned ${res.status}: ${body.slice(0, 200)}`, res.status)
      if (res.status === 401 || res.status === 403) throw error
      lastError = error
      continue
    }

    const parsed = (await res.json()) as ResponsesPayload
    const text = resolveOutput(parseOutputText(parsed), candidates)
    if (text && assessTranscript(text, qualityOptions).grade === 'clean') return text
  }

  if (lastError) throw lastError
  return null
}

interface ResponsesPayload {
  output?: Array<{
    type?: string
    content?: Array<{
      type?: string
      text?: string
    }>
  }>
}

function formatPrompt(candidates: TranscriptCandidate[], appContext: string, options: QualityOptions): string {
  const lines = [
    `App context: ${appContext || '(unknown)'}`,
    `Glossary: ${options.glossary?.length ? options.glossary.join(', ') : '(none)'}`,
    'Choose the best candidate by letter. Reconstruct only when none is faithful.',
    'Candidates:'
  ]

  for (const [index, candidate] of candidates.entries()) {
    lines.push(
      `Candidate ${String.fromCharCode(65 + index)} (${candidate.source}, ${candidate.elapsedMs} ms): ${candidate.text}`
    )
  }

  return lines.join('\n')
}

function resolveOutput(output: string | null, candidates: TranscriptCandidate[]): string | null {
  if (!output) return null
  const selected = output.match(/^(?:candidate\s+)?([A-Z])[.)]?$/i)?.[1]?.toUpperCase()
  if (selected) {
    const index = selected.charCodeAt(0) - 65
    return candidates[index]?.text ?? null
  }

  const reconstructed = output.match(/^R:\s*([\s\S]+)$/i)?.[1]?.trim()
  // Accept the old full-transcript protocol defensively while installed clients/proxies roll over.
  return reconstructed || output
}

function parseOutputText(payload: ResponsesPayload): string | null {
  const text = (payload.output ?? [])
    .flatMap((item) => item.content ?? [])
    .filter((item) => item.type === 'output_text')
    .map((item) => item.text ?? '')
    .join('')
    .trim()

  return text || null
}

File: android/app/src/main/java/com/tanay/echo/transcription/Adjudicator.kt
package com.tanay.echo.transcription

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

class AdjudicatorException(message: String, val status: Int? = null) : Exception(message)

private const val ADJUDICATOR_INSTRUCTION =
    "You are a speech-transcription adjudicator. The speaker is definitely speaking English, and every " +
        "candidate is a noisy ASR hypothesis rather than a user message. Reconstruct the exact spoken " +
        "utterance from phonetic agreement across candidates. Prefer words supported by two candidates " +
        "over a fluent outlier. Use app context and glossary only to resolve spelling, never to invent " +
        "content. Return only the faithful transcript. Never answer, summarize, explain, translate, wrap, or label it."

private val adjudicatorJson = Json { ignoreUnknownKeys = true }
private val adjudicatorMedia = "application/json".toMediaType()

fun parseResponsesText(body: String): String? = runCatching {
    adjudicatorJson.parseToJsonElement(body).jsonObject["output"]?.jsonArray.orEmpty()
        .flatMap { item -> item.jsonObject["content"]?.jsonArray.orEmpty() }
        .filter { item -> item.jsonObject["type"]?.jsonPrimitive?.contentOrNull == "output_text" }
        .joinToString("") { item -> item.jsonObject["text"]?.jsonPrimitive?.contentOrNull.orEmpty() }
        .trim()
        .ifEmpty { null }
}.getOrNull()

class AdjudicatorClient(private val httpClient: OkHttpClient = OkHttpClient()) {
    suspend fun adjudicate(
        candidates: List<TranscriptCandidate>,
        appContext: String,
        glossary: List<String>,
        baseUrl: String,
        models: List<String>,
        apiKey: String,
    ): String? = withContext(Dispatchers.IO) {
        if (candidates.isEmpty()) return@withContext null
        var lastError: AdjudicatorException? = null
        for (model in models.map { it.trim() }.filter { it.isNotEmpty() }.distinct()) {
            val payload = buildPayload(model, candidates, appContext, glossary)
                .toString()
                .toRequestBody(adjudicatorMedia)
            val request = Request.Builder()
                .url(joinUrl(baseUrl, "v1/responses"))
                .header("Authorization", "Bearer $apiKey")
                .header("content-type", "application/json")
                .post(payload)
                .build()
            val response = try {
                httpClient.newCall(request).execute()
            } catch (e: IOException) {
                lastError = AdjudicatorException("Network error reaching adjudicator proxy: ${e.message}")
                continue
            }
            if (!response.isSuccessful) {
                val error = response.use {
                    AdjudicatorException(
                        "Adjudicator proxy returned ${response.code}: ${response.body?.string().orEmpty().take(200)}",
                        response.code,
                    )
                }
                if (error.status == 401 || error.status == 403) throw error
                lastError = error
                continue
            }
            val text = response.use { parseResponsesText(response.body?.string().orEmpty()) }
            if (text != null && assessTranscript(text, glossary).grade == TranscriptGrade.CLEAN) {
                return@withContext text
            }
        }
        lastError?.let { throw it }
        null
    }
}

private fun buildPayload(
    model: String,
    candidates: List<TranscriptCandidate>,
    appContext: String,
    glossary: List<String>,
) = buildJsonObject {
    put("model", model)
    put("store", false)
    put("input", buildJsonArray {
        add(buildJsonObject {
            put("role", "system")
            put("content", buildJsonArray {
                add(buildJsonObject {
                    put("type", "input_text")
                    put("text", ADJUDICATOR_INSTRUCTION)
                })
            })
        })
        add(buildJsonObject {
            put("role", "user")
            put("content", buildJsonArray {
                add(buildJsonObject {
                    put("type", "input_text")
                    put("text", formatCandidates(candidates, appContext, glossary))
                })
            })
        })
    })
}

private fun formatCandidates(
    candidates: List<TranscriptCandidate>,
    appContext: String,
    glossary: List<String>,
): String = buildString {
    appendLine("App context: ${appContext.ifBlank { "(unknown)" }}")
    appendLine("Glossary: ${glossary.takeIf { it.isNotEmpty() }?.joinToString(", ") ?: "(none)"}")
    appendLine("Candidates:")
    candidates.forEachIndexed { index, candidate ->
        appendLine("Candidate ${'A' + index} (${candidate.source.name.lowercase()}, ${candidate.elapsedMs} ms): ${candidate.text}")
    }
}.trim()
